test/libevmasm/PlainAssemblyParser.cpp

```diff
@@ -0,0 +1,182 @@
+			}
+			else
+			{
+				std::string_view immediateArgument = expectArgument();
+				expectNoMoreArguments();
+
+				if (!immediateArgument.starts_with("0x"))
+					BOOST_THROW_EXCEPTION(std::runtime_error(formatError("The immediate argument to PUSH must be a hex number prefixed with '0x'.")));
+
+				immediateArgument.remove_prefix("0x"s.size());
+				codeJSON.push_back({{"name", "PUSH"}, {"value", immediateArgument}});
+			}
+		}
+		else if (currentToken().value == "tag")
+		{
+			std::string_view tagID = expectArgument();
+			expectNoMoreArguments();
+
+			codeJSON.push_back({{"name", "tag"}, {"value", tagID}});
+			codeJSON.push_back({{"name", "JUMPDEST"}});
+		}
+		else
+			BOOST_THROW_EXCEPTION(std::runtime_error(formatError("Unknown instruction.")));
+	}
+	return {{".code", codeJSON}};
+}
+
+PlainAssemblyParser::Token const& PlainAssemblyParser::currentToken() const
+{
+	soltestAssert(m_tokenIndex < m_lineTokens.size());
+	return m_lineTokens[m_tokenIndex];
+}
+
+PlainAssemblyParser::Token const& PlainAssemblyParser::nextToken() const
+{
+	soltestAssert(m_tokenIndex + 1 < m_lineTokens.size());
+	return m_lineTokens[m_tokenIndex + 1];
+}
+
+bool PlainAssemblyParser::advanceToken()
+{
+	if (!hasMoreTokens())
+		return false;
+
+	++m_tokenIndex;
+	return true;
+}
+
+std::string_view PlainAssemblyParser::expectArgument()
+{
+	bool hasArgument = advanceToken();
+	if (!hasArgument)
+		BOOST_THROW_EXCEPTION(std::runtime_error(formatError("Missing argument(s).")));
+
+	return currentToken().value;
+}
+
+void PlainAssemblyParser::expectNoMoreArguments()
+{
+	bool hasArgument = advanceToken();
+	if (hasArgument)
+		BOOST_THROW_EXCEPTION(std::runtime_error(formatError("Too many arguments.")));
+}
+
+void PlainAssemblyParser::advanceLine(std::string_view _line)
+{
+	++m_lineNumber;
+	m_line = _line;
+	m_lineTokens = tokenizeLine(m_line);
+	m_tokenIndex = 0;
+}
+
+std::vector<PlainAssemblyParser::Token> PlainAssemblyParser::tokenizeLine(std::string_view _line)
+{
+	auto const notWhiteSpace = [](char _c) { return !isWhiteSpace(_c); };
+
+	std::vector<Token> tokens;
+	auto tokenLocation = boost::find_token(_line, notWhiteSpace, boost::token_compress_on);
+	while (!tokenLocation.empty())
+	{
+		std::string_view value{tokenLocation.begin(), tokenLocation.end()};
+		if (value.starts_with("//"))
+			break;
+
+		tokens.push_back({
+			.value = value,
+			.position = static_cast<size_t>(std::distance(_line.begin(), tokenLocation.begin())),
+		});
+		soltestAssert(!value.empty());
+		soltestAssert(tokens.back().position < _line.size());
+		soltestAssert(tokens.back().position + value.size() <= _line.size());
+
+		std::string_view tail{tokenLocation.end(), _line.end()};
+		tokenLocation = boost::find_token(tail, notWhiteSpace, boost::token_compress_on);
+	}
+
+	return tokens;
+}
+
+std::string PlainAssemblyParser::formatError(std::string_view _message) const
+{
+	soltestAssert(currentToken().value.size() >= 1);
```

Comment:
I found it a bit odd to only assert this here, perhaps this should be rather treated as an invariant of tokens in general and be added to `tokenizeLine` - or even to a ctor of `Token` (although I do understand that it is a nice POD right now and that would destroy that property).

Reply:
Looking at it now, I actually don't strictly need this assert. I removed it now.

I don't remember why exactly I needed it, but it was not a general requirement, rather something local to `formatError()`. I initially stored both `start` and `end` of the token in the struct and I think that something in the formatting logic would break if they were equal. In any case, that's no longer relevant.